Fix deriving accessors in uncurried mode

JonoPrest opened the pull request:
Closes #6492

Uncurried mode breaks `deriving(accessors)`

This PR converts the derived accessor functions into uncurried functions if Config is "Uncurried".

Previously (example from test added) 

```res
@@uncurried
@deriving(accessors)
type variant = | Num(int) 

//Asserts the correct signature for derived accessor
let _numAlias: int => variant = num

//Asserts that inference works when composing
//with derived functions
let compose = (a, accessor) => accessor(a)
let _composedNum = compose(1, num)

```

would error as below
```sh
  4 │
  5 │ //Asserts the correct signature for derived accessor
  6 │ let _numAlias: int => variant = num
  7 │
  8 │ //Asserts that inference works when composing

  This function is a curried function where an uncurried function is expected
```

zth commented:
Impressive! Ping @cristianoc who's best to review this.

zth commented:
Should probably ship this in 11.1. Mind changing the base branch to that? 11.0_release

cristianoc commented on jscomp/build_tests/uncurried_accessors/input.js:
```diff
@@ -0,0 +1,15 @@
+var p = require("child_process");
```
I think a build test is not required for this, as the issue shows with some simple files that don't compile.
Adding a file to `jscomp/test` should be enough, as compile errors would be picked up.

cristianoc replied:
Plus the compiled output would be checked in.

JonoPrest replied:
Ok great thanks for the feedback. I'll do that 👍🏼

JonoPrest commented:
Apologies @cristianoc made a mistake with the rebase if you are mid review, busy fixing now.

JonoPrest commented:
> Impressive! Ping @cristianoc who's best to review this.

Thanks @zth! Just rebased and I'll implement the changes suggested by Cristiano 👍🏼

JonoPrest commented:
Hey @cristianoc, 

Note I've moved out of the tests out of the build test folder. I also caught a bug realising I was not checking for Variants without parameters.   So I've made a tweak to the derive projector file and added a test case for this.

The reason I initially had it as a build test was because I wanted to show in a commit that it was failing and have a passing test that asserts the failure. Then in the following commit that fixes the behaviour invert the assertion that shows its fixed as I like reading git history that tells a story like that 😄 

Out of interest is there a nice simple way to assert some sort of failure?

cristianoc reviewed (approved):
Looks great

zth commented:
Very well done @JonoPrest!

JonoPrest commented:
> Very well done @JonoPrest!

Thanks @zth and thanks for the review @cristianoc, very grateful for your dedication to rescript and openness to contribution 🙏🏼
